tuna/miopen/fin_builder.py

```diff
@@ -85,18 +86,37 @@ def process_pdb_compile(self, session, fin_json):
 
     return status
 
+  def close_job(self):
+    """mark a job complete"""
+    self.set_job_state('compiled')
+
+  def result_queue_drain(self):
+    """check for lock and commit the result queue"""
+    if self.result_queue_lock.acquire(block=False):
+      with DbSession() as session:
+        self.result_queue_commit(session, self.close_job)
+      self.result_queue_lock.release()
+      return True
+    return False
+
   def step(self):
     """Main functionality of the builder class. It picks up jobs in new state and compiles them"""
+    self.pending = []
+    self.result_queue_drain()
 
     # pylint: disable=duplicate-code
-    try:
-      self.check_env()
-    except ValueError as verr:
-      self.logger.error(verr)
-      return False
+    if self.first_pass:
+      self.first_pass = False
+      try:
+        self.check_env()
+      except ValueError as verr:
+        self.logger.error(verr)
+        return False
     # pylint: enable=duplicate-code
 
     if not self.get_job("new", "compile_start", True):
+      while not self.result_queue_drain():
+        sleep(1)
```

Comment:
Please implement a proper back off here. I believe the implementation exists in Tuna for a randomized backoff to prevent lockup

Reply:
sure, I can add the random here.